esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -1241,194 +1241,205 @@ void setDeleteFileResults(const char* fileName, const char* nodeGroup, bool fail
+                userdesc->set(user, context.queryPassword(), context.querySignature());
             }
+
+            auditStr.set(",FileAccess,WsDfu,DELETED,");
+            auditStr.append(espProcess.get());
+            auditStr.append(',');
+            if (!isEmptyString(user))
+                auditStr.append(user).append('@');
+            context.getPeer(auditStr);
         }
-        fdir.removeEntry(fn, userdesc, NULL, REMOVE_FILE_SDS_CONNECT_TIMEOUT, true);
-        LOG(MCauditInfo, "%s,%s", auditStr, fn);
-        setDeleteFileResults(lfn, group, false, isSuper ? "Deleted Superfile" : "Deleted File", NULL, returnStr, actionResults);
-    }
-    catch(IException* e)
-    {
-        StringBuffer emsg;
-        e->errorMessage(emsg);
-        if (removeFromSuperfiles && strstr(emsg, "owned by"))
-        {
-            if (!doRemoveFileFromSuperfiles(lfn, userdesc, superFiles, failedFiles, deleteRecursively, auditStr, returnStr, actionResults))
-                return DeleteActionFailure;
-            return doDeleteFile(fn, userdesc, superFiles, failedFiles, auditStr, returnStr, actionResults, superFilesOnly, false, false);
-        }
-        if (e->errorCode() == DFSERR_CreateAccessDenied)
-            emsg.replaceString("Create ", "Delete ");
 
-        setDeleteFileResults(lfn, group, true, "Could not delete", emsg.str(), returnStr, actionResults);
-        e->Release();
-        return DeleteActionFailure;
-    }
-    catch(...)
-    {
-        setDeleteFileResults(lfn, group, true, "Could not delete", "unknown exception", returnStr, actionResults);
-        return DeleteActionFailure;
-    }
-    return DeleteActionSuccess;
-}
+        const char *getReturnStr() const { return returnStr.str(); }
+        IArrayOf<IEspDFUActionInfo> &getActionResults() { return actionResults; }
 
-void doDeleteFiles(StringArray &files, IUserDescriptor *userdesc, StringArray &superFiles, StringArray &failedFiles,
-    const char *auditStr, StringBuffer &returnStr, IArrayOf<IEspDFUActionInfo> &actionResults,
-    bool superFilesOnly, bool removeFromSuperfiles, bool deleteRecursively)
-{
-    ForEachItemIn(i, files)
-    {
-        const char* fn = files.item(i);
-        if(!fn || !*fn)
-            continue;
-
-        PROGLOG("Deleting %s", fn);
-        if (DeleteActionFailure==doDeleteFile(fn, userdesc, superFiles, failedFiles, auditStr, returnStr, actionResults, superFilesOnly, removeFromSuperfiles, deleteRecursively))
+        void deleteFiles(StringArray &logicalFileNames, bool removeFromSuperFiles, bool removeEmptyOwningSuperFiles)
         {
-            failedFiles.appendUniq(fn);
-            PROGLOG("Delete %s failed", fn);
+            IDistributedFileDirectory &fdir = queryDistributedFileDirectory();
+            cleanOwningSuperFiles(fdir, logicalFileNames, removeFromSuperFiles, removeEmptyOwningSuperFiles);
+            ForEachItemIn(i, logicalFileNames)
+                deleteFile(fdir, logicalFileNames.item(i));
         }
```

Comment:
the code would be further simplified if there was a single loop of logicalFileNames here.
i.e. do the logic of removing the file from owning supers per file as you go.
And also, I think you kind of have to do that (or it makes more sense) from the point of view of checking to remove empty supers.
i.e. still check if super owners are empty as you go and at them to a unique list, if a subsequent file to be deleted is a super and in that list, you can skip doing anything else with it. Then at end when you have done processing all files to be deleted, delete the empty super files. Better to do at end because they may only become empty after several files have been deleted.

Reply:
> Then at end when you have done processing all files to be deleted, delete the empty super files. Better to do at end because they may only become empty after several files have been deleted.

I am re-thinking that. Why cannot delete an empty super file as soon as its numSubFiles() is 0? The emptyOwningSuperFiles list will not be needed if an empty super file is deleted at that time. But, will that cause any problem because the Owned<IDistributedSuperFileIterator> superOwners is still in the scope?

Reply:
>But, will that cause any problem because the Owned superOwners is still in the scope?

yes, and I forgot to mention something related.

It is better (more efficient) to use the IDistributeFile to delete the file, rather than release it, and call queryDistributedFileDirectory().removeEntry.
i.e. use IDistributedFile::detach()    - with a short timeout.

Reply:
> Why cannot delete an empty super file as soon as its numSubFiles() is 0?

You could, but worth maintaining a list, so that when you're looking to delete the next file(s), you can check if already in the list, i.e. already deleted, or to be deleted.

Reply:
> use IDistributedFile::detach() - with a short timeout.

How does it work? When the queryDistributedFileDirectory().removeEntry(fn, ...) is called, the fn may contain group information. If one logical file belongs to 2 groups. I assume that the removeEntry() may only delete a logical file from one group. Will the IDistributedFile::detach() also delete a logical file from one group?